### ISSUE TITLE
can't create new snapshot

### ISSUE THREAD

**Issue by knutov** [borgbackup/borg#2963]
`borg info` is working, `borg check` did not find anything wrong, but `borg create` does not working.

What should I do in this case?

```
# borg info $REPOSITORY
Warning: Attempting to access a previously unknown unencrypted repository!
Do you want to continue? [yN] y
Repository ID: 6762b6ecbfdac6863c5098d7eb8798ef6957038a39a2b057a61d8345e9907e40
Location: /abcd/borg/USER
Encrypted: No
Cache: /root/.cache/borg/6762b6ecbfdac6863c5098d7eb8798ef6957038a39a2b057a61d8345e9907e40
Security dir: /root/.config/borg/security/6762b6ecbfdac6863c5098d7eb8798ef6957038a39a2b057a61d8345e9907e40
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
All archives:               90.78 GB             57.84 GB              6.87 GB

                       Unique chunks         Total chunks
Chunk index:                  101892              8291571
```

borg create:

```
Start backup USER
A repository already exists at bak01@server.tld:borg/USER.
Traceback (most recent call last):

  File "borg/remote.py", line 239, in serve

  File "borg/repository.py", line 1004, in get

  File "borg/repository.py", line 343, in get_transaction_id

  File "borg/repository.py", line 324, in check_transaction

  File "borg/repository.py", line 1150, in get_segments_transaction_id

  File "borg/repository.py", line 1171, in is_committed_segment

PermissionError: [Errno 13] Permission denied: '/abcd/borg/USER/data/0/1138'

Borg server: Platform: Linux server1 2.6.32-042stab120.5 #1 SMP Tue Oct 25 22:31:12 MSK 2016 x86_64 x86_64
Borg server: Linux: CentOS 6.9 Final
Borg server: Borg: 1.1.0rc1  Python: CPython 3.5.3
Borg server: PID: 790687  CWD: /abcd
Borg server: sys.argv: ['borg', 'serve', '--umask=077', '--info']
Borg server: SSH_ORIGINAL_COMMAND: None
Platform: Linux jane 2.6.32-042stab123.1 #1 SMP Wed Mar 22 15:21:30 MSK 2017 x86_64 x86_64
Linux: debian jessie/sid
Borg: 1.1.0rc1  Python: CPython 3.5.3
PID: 24661  CWD: /home/USER
sys.argv: ['borg', 'create', '--compression', 'lz4', '-svpx', '--ignore-inode', '--show-rc', '--exclude', '*bitrix/updates/*', '--exclude', '*bitrix/*cache/*', '--exclude', '*includes/sessions/sess*', '--exclude', '*livestreet_cache*', '--exclude-caches', '--exclude-if-present', '.nobackup', 'bak01@server.tld:borg/USER::USER-{now:%Y-%m-%d-%H%M%S}', './']
SSH_ORIGINAL_COMMAND: None

terminating with error status, rc 2
```

first snapshots was created with borg 1.0.10

now borg 1.1.0rc1 from both side (client and server).

**upd:**
About `Permission denied: '/abcd/borg/USER/data/0/1138'` - there was no sufficiant disk space I suppose, and the file now 

```
ll /abcd/borg/USER/data/0/1138
-rw------- 1 root root 17 Aug 18 06:55 /vz/bak01/borg/a2892/data/0/1138
```
I think borg should fix this situation automatically. Any way, how to fix it now? Is it right to delete this file and create new snapshot?

**Comment by ThomasWaldmann** [comment]
If you have permission issues, it is likely because you once ran as root and now you run as other user.

Just always use same user.

**Comment by knutov** [comment]
only `borg check` was run as root on the server side. Can it be the reason in this case?

all `borg create` are done from client side and uses remote repo with non-root user.

**Comment by ThomasWaldmann** [comment]
Yes, always run as same user. If you don't borg will create files as root and later the normal user can't read them.

**Comment by knutov** [comment]
Well, I fixed file owner. Now `borg create` just hangs after

```
Start backup USER
A repository already exists at bak01@***.net:borg/USER.`
```

**Comment by ThomasWaldmann** [comment]
are you running `borg init` onto a already inited repo location? don't, you only need init once.

**Comment by knutov** [comment]
Yes.

In my case it is hard to check if it is created already, so i do `borg init ... ; borg create ...`.

This should not lead to any effects in this case, `borg init` just terminates with the error and next `borg create` executed.

The real problem is:

1. the disk was ended
1. now `borg create` hangs

**Comment by ThomasWaldmann** [comment]
You need free space, borg can't work without.

**Comment by knutov** [comment]
but now I have free space. And `borg create` does nothing.

upd: Oh, I found, after ctrl+c this time it showed

```
Traceback (most recent call last):
  File "borg/locking.py", line 129, in acquire
FileExistsError: [Errno 17] File exists: '/root/.cache/borg/6762b6ecbfdac6863c5098d7eb8798ef6957038a39a2b057a61d8345e9907e40/lock.exclusive'
```

So there was some old (month+ ago) lock that was not autodetected.

I deleted and fetched again the lock and the cache and now everything is ok.

**Comment by ThomasWaldmann** [comment]
OK, so guess nothing left here.

**Comment by knutov** [comment]
Well, now I'm sure, It must react in some different way.

`borg create` silently (!!!) hangs and doing nothing.

Wtih --debug I see

```
Remote: using builtin fallback logging configuration
Remote: 35 self tests completed in 0.11 seconds
Remote: using builtin fallback logging configuration
Remote: Initialized logging system for new (v3) protocol
Remote: Resolving repository path b'borg/b2290'
Remote: Resolved repository path to '/vz/bak01/borg/b2290'
TAM-verified manifest
```

That's all.

Ok, I can manually interrupt it with ctrl+c and now I see 

```
RemoteRepository: 190 B bytes sent, 2.47 kB bytes received, 3 messages sent
Connection closed by remote host
Traceback (most recent call last):
  File "borg/locking.py", line 129, in acquire
FileExistsError: [Errno 17] File exists: '/root/.cache/borg/0a9e7112cc6c94f0a345a7f970fc4c7b606fb0e6c5cc84639cc9fceb261bbc6d/lock.exclusive'
```

Why there is no timeout that will halt in this case?

For example, i have shell script which should backup some folders to different repos (for example /home/* , each user to different repo). I do it in a cycle, and some user infinitely broke all backup process. This is awfull!

And - all locks were broken, why it was not detected and fixed automatically?

Or may be there is some way to control it? I did not find it in documentation.

**Comment by ThomasWaldmann** [comment]
There is a quite short default timeout for acquiring the lock (few seconds).

That acquire() of the locking systems finds an existing lock directory just means there currently is a lock.

If you hit ctrl-c within the --lock-wait time, that "File exists" might be seen in a traceback, but it by itself is not an error condition.

If you think there is something wrong with the locking, create a minimal reproducable sequence of commands that reproduce the problem.

**Comment by knutov** [comment]
and default timeout is 1 second, I know. But does it work?

I hit ctrl+c much after default timeout.

Now I have lock which mention pid that is not exists.

I suppose borg must check pid (in my case it's local) and automatically do break-lock in case pid does not exist.

It's hardly to reproduce in nature, I have thousands of repos and sometimes it happens.
The case to repeat, I suppose, is to run borg create and kill -9 (In my case I might had OOM). 
I repeatedly run in this situation after switch from 1.0.10 to 1.1.0rc1/rc2.

**Comment by ThomasWaldmann** [comment]
automatically breaking of (definitely sure) stale locks will come in borg 1.1. of course we can only be sure for locks created by the local host, because we can't know whether a process on another host is still alive or not.

if you (or the kernel OOM killer) shoots the borg process, there will be a lock left behind, naturally.
borg on same host later should be able to clean that up, but not from another host.

**Comment by knutov** [comment]
in this particular case lock are local (host of pid is the same).

and another important problem I mentioned now and before - borg silently hangs. So there is no timeouts at all (I found one server with runned borgs more the month ago).

silently hangs in case of any problem - locks, wrong owners of files from serverside, any problems. It will show the problem in case of --debug and ctrl+c, but backup is crontab job, it suppose to be no alive man to check and control it. I don't  think this all is good way.

**Comment by knutov** [comment]
so, there is no `borg check-lock` and `--lock-wait N` does not work in this case (I set --lock-wait 1 and it waits infinitely), so how to fix it?

**Comment by ThomasWaldmann** [comment]
```
(borg-env) tw@tux:~/w/borg$ borg create -v --progress repo::foo2 /usr/share
Killed  MB O 51.31 MB C 48.69 MB D 1694 N usr/share/qt4/mkspecs/aix-xlc                                                   
# ^^^ I did killall -9 borg from other terminal.

(borg-env) tw@tux:~/w/borg$ borg create -v --progress repo::foo3 /usr/share
Killed stale lock tux@180546901543488.7477-0.
Removed stale exclusive roster lock for pid 7477.
Removed stale exclusive roster lock for pid 7477.
Killed stale lock tux@180546901543488.7477-0.
Removed stale exclusive roster lock for pid 7477.
Removed stale exclusive roster lock for pid 7477.
C 54.73 MB D 2922 N usr/share/npm/node_modules/request/no...ode_modules/readable-stream/transform.js
...
```
(

**Comment by knutov** [comment]
I saw this behavior ones too. I thought it will always work.

Ok, why can it not work in my environment? What should I check?

**Comment by ThomasWaldmann** [comment]
same for "remote" repo:
```
(borg-env) tw@tux:~/w/borg$ borg create -v --progress --remote-path=borg-dev ssh://tw@localhost/home/tw/w/borg/repo::foo4 /usr/share
Killed   MB O 57.86 MB C 55.11 MB D 3047 N usr/share/npm/node_modules/request/no...ules/hawk/node_modules/boom/lib/index.js
# ^^^ did killall -9 borg

(borg-env) tw@tux:~/w/borg$ borg create -v --progress --remote-path=borg-dev ssh://tw@localhost/home/tw/w/borg/repo::foo5 /usr/share
Killed stale lock tux@180546901543488.7576-0.
Removed stale exclusive roster lock for pid 7576.
Removed stale exclusive roster lock for pid 7576.
^CConnection closed by remote host691 N usr/share/qt4/mkspecs/win32-msvc2010/qplatformdefs.h                            
# hit ctrl-c after watching progress for a while
```

**Comment by ThomasWaldmann** [comment]
not sure. first i'ld check that all files/dirs are owned by correct uid:
- for the client(s) this is whatever user you run borg under (maybe root).
- for the server, this is always the same single user (usually not root) for all the clients.

check:
- repo files
- ~/.config/borg/...
- ~/.cache/borg/...

**Comment by knutov** [comment]
I mostly do not know python, but I looked https://github.com/borgbackup/borg/blob/master/src/borg/locking.py and it looks fine.

owners ar definitely right now.

and lock is:

```
root@bx5:~/.cache/borg/a7a5b69715cf2b6190e78634deb432d04f12b3b983fb5fbac5db93f4aa8416e5/lock.exclusive# ll
total 8
drwx------ 2 root root 4096 Jun 28 03:37 ./
drwx------ 5 root root 4096 Jun 28 03:37 ../
-rw------- 1 root root    0 Jun 28 03:37 bx5.3885-0
root@bx5:~/.cache/borg/a7a5b69715cf2b6190e78634deb432d04f12b3b983fb5fbac5db93f4aa8416e5/lock.exclusive# ps 3885
  PID TTY      STAT   TIME COMMAND
root@bx5:~/.cache/borg/a7a5b69715cf2b6190e78634deb432d04f12b3b983fb5fbac5db93f4aa8416e5/lock.exclusive#
```

**Comment by knutov** [comment]
If I do manually `borg break-lock` it will start work, I checked with another repo.

**Comment by ThomasWaldmann** [comment]
btw, there is one lock for the cache and one for the repo.

**Comment by knutov** [comment]
this lock is what borg said after ctrl+c

**Comment by knutov** [comment]
you can see - there is no process with this pid, and lock was created month+ ago, and it was not removed automatically

**Comment by ThomasWaldmann** [comment]
ehrm, did you have the cwd inside the lock dir while borg was running?

**Comment by knutov** [comment]
no, just after to show content of the folder. And I'm sure no other sessions or processes are inside.

**Comment by ThomasWaldmann** [comment]
for 1.1, the lock name should be `<socket.getfqdn()>`@`<uuid.getnode()`, but I don't see an @ there.

so, are we talking about 1.0.x or 1.1?

**Comment by knutov** [comment]
oh, can the root of problem be in that lock was stale with 1.0 version and now 1.1 expects new format but gets old and can't deal with it?

**Comment by ThomasWaldmann** [comment]
yes, could be. guess one should not upgrade borg with locked repos.

**Comment by ThomasWaldmann** [comment]
(we had to add that @node part to make rather sure that the id is unique, e.g. for people who have multiple hosts named "server")

**Comment by knutov** [comment]
yes, format is different, now I have file

`lock.exclusive/bx5.3885-0`
and file `lock.roster` contained `{"exclusive": [["bx5", 3885, 0]]}`.

ok, now I'll do breake-lock.

so, two problems left:

1) borg silently hangs in case of unexpected errors
2) lock-timeout does not work in this case
3) and why parser of lock file says nothing about wrong format? )

**Comment by ThomasWaldmann** [comment]
#3014

**Comment by ThomasWaldmann** [comment]
again: if you want a bug fixed, provide a way to reproduce it (and create one issue per different bug).

the parser treats the stuff left of the pid as "id" and does not parse it further.

**Comment by knutov** [comment]
ok, 

> people who have stale locks from before upgrade will just have to use break-lock, as they also did in 1.0 when running into a leftover lock.

People should to know about their problem somehow. Now with any of mentioned in this issue problems there are no timeouts and borg silently hang, until manual interruption.

What I expect as user is borg exit with some error text. Even text like "something strange happend, run with --debug" is good, but borg shoud exit with error.

Then automated script can read exit code and other magic will happen, but the big problem now is  borg just infinitely wait.

**Comment by knutov** [comment]
reproduce way for this can be manually creating lock file in old format

**Comment by ThomasWaldmann** [comment]
```
borg-env) tw@tux:~/w/borg$ mkdir repo/lock.exclusive
(borg-env) tw@tux:~/w/borg$ touch repo/lock.exclusive/tux.123456-0
(borg-env) tw@tux:~/w/borg$ borg list repo
Failed to create/acquire the lock /home/tw/w/borg/repo/lock.exclusive (timeout).
```

So, still waiting for your script to reproduce the issue...

**Comment by knutov** [comment]
fantastic.

I saved files (but not all repop cache) before break-lock. The only thing was unsaved is dir with uncommited transactions.

 what does exactly `break-lock` except deleting those files?

**Comment by ThomasWaldmann** [comment]
break-lock unconditionally kills the exclusive lock and also the lock roster.

**Comment by ThomasWaldmann** [comment]
btw, apropos uncommitted transactions: i just fixed a bug relating to them, see #3002.
